staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go

```diff
@@ -414,9 +432,16 @@ func (w *watchCache) GetAllEventsSinceThreadUnsafe(resourceVersion uint64) ([]*w
 	size := w.endIndex - w.startIndex
 	// if we have no watch events in our cache, the oldest one we can successfully deliver to a watcher
 	// is the *next* event we'll receive, which will be at least one greater than our current resourceVersion
-	oldest := w.resourceVersion + 1
+	oldest := w.resourceVersion
 	if size > 0 {
-		oldest = w.cache[w.startIndex%w.capacity].resourceVersion
+		// if the first item is a marker, we require version to be equal at least to it.
+		// Otherwise, it's fine to have it lower by one.
+		item := w.cache[w.startIndex%w.capacity]
+		if item.watchCacheEvent.isMarker() {
```

Comment:
In my personal opinion this is a bit less clear, but yeah - we can do that. I can change that if you prefer this approach (the end result is the same in both cases).

Reply:
I like keeping separate data separate, rather than adding a synthetic watch event to the buffer and then having to skip sending it. I'd also like to keep the oldest variable meaning "the resourceVersion of the oldest watch event we can deliver" (renaming the variable to make that clearer is fine)

Reply:
Fair enough.
The code is now changed - the test I added remained untouched.